app/javascript/packages/document-capture/components/form-error-message.jsx

```diff
@@ -0,0 +1,26 @@
+import React from 'react';
+import { RequiredValueMissingError } from './form-steps';
+import useI18n from '../hooks/use-i18n';
+
+/** @typedef {import('react').ReactNode} ReactNode */
+
+/**
+ * @typedef FormErrorMessageProps
+ *
+ * @prop {Error} error Error for which message should be generated.
+ */
+
+/**
+ * @param {FormErrorMessageProps} props Props object.
+ */
+function FormErrorMessage({ error }) {
+  const { t } = useI18n();
+
+  if (error instanceof RequiredValueMissingError) {
+    return <>{t('simple_form.required.text')}</>;
+  }
```

Comment:
I bounced around between a few ideas, but a common underlying challenge is knowing where we can have access to translated texts. In most of the application, we rely on what's provided through the top-level React context, using `useI18n`, which in turn uses `useContext(I18nContext)`.

But that really only works from a component, and not from a place like `DocumentsStep.validate` where the errors are constructed, since it's more of a standard function where we can't (or at least shouldn't) rely on React hooks and context.

My general approach with this was similar to what you and I were thinking with upload error codes sent by the server being mapped to corresponding messages where they're used in the UI. In this case, the error codes are represented by custom `Error` classes.

Where exactly those classes are converted into localized strings is flexible, but ideally it ideally needs to integrate seamlessly with how AcuantCapture and FileInput manage their own errors. It dawned on me with something like `FileInput`'s `error` prop, while often represented by a `string`, in general usage could need to support any react node (for example, to be able to include a link in the error message).

Choosing to generate the error string here was largely a result of being a component which has access to React context (`useI18n`). And while it's not quite as close as the error itself, it's at least centralized enough to avoid a case that each of `AcuantCapture` and `FileInput` would not only need to map `instanceof Error` to a message, but also juggle multiple possible types of errors (its own `string` errors and `Error` types).

Reply:
There might be an alternative implementation where validation happens _in_ a component which has access to the React context, in which case something like what you're proposing could be more feasible, where either the `Error` classes can be constructed with the localized message, or maybe we don't need `Error` instances at all (just use the error string directly).

The problem is that this validation needs to be able to be triggered externally, by `FormSteps`, in response to a submission, which really isn't part of the behavior of a step. We might be able to get `FormSteps` to signal to a component to validate its own inputs, but `FormSteps` still needs to know whether to allow the user to proceed to the next step. On top of that, there's the whole "allow history to work as expected, but also validate after refresh".